File: exercises/week02/di/src/decoupledwithfactory/MessageSupportFactory.java
package decoupledwithfactory;

import java.io.FileInputStream;
import java.util.Properties;

public class MessageSupportFactory {

    private static MessageSupportFactory instance = null;

    static {
        instance = new MessageSupportFactory();
    }

    private Properties props = null;
    private MessageRenderer renderer = null;
    private MessageProvider provider = null;

    private MessageSupportFactory() {
        props = new Properties();

        try {
            props.load(new FileInputStream("/Users/keith/Courses/sdp/2016/SDP2016/di/src/decoupledwithfactory/bean.properties"));

            // get the implementation classes
            String rendererClass = props.getProperty("renderer.class");
            String providerClass = props.getProperty("provider.class");

            renderer = (MessageRenderer) Class.forName(rendererClass)
                    .newInstance();
            provider = (MessageProvider) Class.forName(providerClass)
                    .newInstance();
        } catch (Exception ex) {
            System.err.println("Failed to create factory methods");
            ex.printStackTrace();
        }
    }

    public static MessageSupportFactory getInstance() {
        return instance;
    }

    public MessageRenderer getMessageRenderer() {
        return renderer;
    }

    public MessageProvider getMessageProvider() {
        return provider;
    }

}

File: exercises/week02/di-scala/src/decoupledwithfactory/MessageSupportFactory.scala
package decoupledwithfactory

import java.io.FileInputStream
import java.util.Properties

object MessageSupportFactory {
  private val instance: MessageSupportFactory = new MessageSupportFactory

  def getInstance: MessageSupportFactory = {
    instance
  }
}

class MessageSupportFactory private() {
  // The constructor is marked private in the class declaration

  private var props: Properties = null
  private var renderer: MessageRenderer = null
  private var provider: MessageProvider = null

  props = new Properties
  try {
    props.load(new FileInputStream("/Users/keith/CloudDocs/Courses/sdp/2016/SDP2016/di-scala/src/decoupledwithfactory/bean.properties"))
    val rendererClass: String = props.getProperty("renderer.class")
    val providerClass: String = props.getProperty("provider.class")
    renderer = Class.forName(rendererClass).newInstance.asInstanceOf[MessageRenderer]
    provider = Class.forName(providerClass).newInstance.asInstanceOf[MessageProvider]
  }
  catch {
    case ex: Exception => {
      System.err.println("Failed to create factory methods")
      ex.printStackTrace
    }
  }

  def getMessageRenderer: MessageRenderer = {
    renderer
  }

  def getMessageProvider: MessageProvider = {
    provider
  }
}